### ISSUE TITLE
Reify "DebugSession"

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode#48377]
Today the VS Code debugger implementation has no object that tracks the full life-cycle of a debug session (including multiple restarts).

The lack of a DebugSession makes it difficult to solve issues related to the lifecycle of a debug session.

**Comment by isidorn** [comment]
fixed via cd1969b4eae58c02ed7a48c212a6f278a10599fc and 9eef194c04

Note that explicti user restarts still ends the DebugSession and creates a new one with a new id, however the re-attach now nicely keeps the session.
Restart is a more heavy operation since we do a [lot of things on top](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugService.ts#L1080)
If you feel strong that a DebugSession should survive a restart we can also change that.

**Comment by weinand** [comment]
I'm reopening this and after two days of debugging issue #55955 I'm sure that we have to really fix this.

**Comment by isidorn** [comment]
I have refactored the code in such a way that the `DebugSession` now encapsulates the creation of the `rawDebugSession` and handling of all the listeners associated with it. imho the code structure is now much nicer as we have got rid of all the ugly maps and confusion in `DebugService`

I have tried to reproduce the EH reload issue and could not (both out of source and with a locally build vscode).

This refactoring has been large so we shuold keep our eyes open for potential regressions.

Also note that some of the events that were living in `DebugService` now live on the `DebugSession`, examples being `onDidCustomEvent` and `onDidLoadedSource`. Due to that I have touched a bit the `loadedScriptsView`.

Feedback very welcome
